exchange/bitswap/workers.go

```diff
@@ -89,27 +72,45 @@ func (bs *Bitswap) taskWorker(ctx context.Context, id int) {
 	}
 }
 
-func (bs *Bitswap) provideWorker(ctx context.Context, id int) {
-	idmap := eventlog.LoggableMap{"ID": id}
-	for {
-		log.Event(ctx, "Bitswap.ProvideWorker.Loop", idmap)
-		select {
-		case k, ok := <-bs.provideKeys:
-			log.Event(ctx, "Bitswap.ProvideWorker.Work", idmap, &k)
-			if !ok {
-				log.Debug("provideKeys channel closed")
-				return
-			}
-			ctx, cancel := context.WithTimeout(ctx, provideTimeout)
-			err := bs.network.Provide(ctx, k)
-			if err != nil {
+func (bs *Bitswap) provideWorker(px process.Process) {
+
+	limiter := ratelimit.NewRateLimiter(px, provideWorkerMax)
+
+	limitedGoProvide := func(k key.Key, wid int) {
+		ev := eventlog.LoggableMap{"ID": wid}
+		limiter.LimitedGo(func(px process.Process) {
+
+			ctx := waitable.Context(px) // derive ctx from px
+			defer log.EventBegin(ctx, "Bitswap.ProvideWorker.Work", ev, &k).Done()
+
+			ctx, cancel := context.WithTimeout(ctx, provideTimeout) // timeout ctx
+			defer cancel()
+
+			if err := bs.network.Provide(ctx, k); err != nil {
 				log.Error(err)
 			}
-			cancel()
-		case <-ctx.Done():
-			return
-		}
+		})
 	}
+
+	// worker spawner, reads from bs.provideKeys until it closes, spawning a
+	// _ratelimited_ number of workers to handle each key.
+	limiter.Go(func(px process.Process) {
```

Comment:
No, this is on the limiter to make sure the limiter doesn't close while we keep adding things to it. it's a bad api-- it could use improving. dont have concrete good ideas on it yet